src/fat.rs

```diff
@@ -353,47 +356,22 @@ impl<'a> OnDiskDirEntry<'a> {
     }
 
     /// If this is an LFN, get the contents so we can re-assemble the filename.
-    pub fn lfn_contents(&self) -> Option<(bool, u8, [char; 13])> {
+    pub fn lfn_contents(&self) -> Option<(bool, u8, String<U26>)> {
         if self.is_lfn() {
-            let mut buffer = [' '; 13];
             let is_start = (self.data[0] & 0x40) != 0;
             let sequence = self.data[0] & 0x1F;
-            // LFNs store UCS-2, so we can map from 16-bit char to 32-bit char without problem.
-            buffer[0] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[1..=2]))).unwrap();
-            buffer[1] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[3..=4]))).unwrap();
-            buffer[2] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[5..=6]))).unwrap();
-            buffer[3] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[7..=8]))).unwrap();
-            buffer[4] = core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[9..=10])))
-                .unwrap();
-            buffer[5] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[14..=15])))
-                    .unwrap();
-            buffer[6] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[16..=17])))
-                    .unwrap();
-            buffer[7] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[18..=19])))
-                    .unwrap();
-            buffer[8] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[20..=21])))
-                    .unwrap();
-            buffer[9] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[22..=23])))
-                    .unwrap();
-            buffer[10] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[24..=25])))
-                    .unwrap();
-            buffer[11] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[28..=29])))
-                    .unwrap();
-            buffer[12] =
-                core::char::from_u32(u32::from(LittleEndian::read_u16(&self.data[30..=31])))
-                    .unwrap();
-            Some((is_start, sequence, buffer))
+            let mut buffer = [0u8; 26];
+            buffer[..10].copy_from_slice(&self.data[1..=10]);
+            buffer[10..22].copy_from_slice(&self.data[14..=25]);
+            buffer[22..].copy_from_slice(&self.data[28..=31]);
+            let data: &[u16; 13] = unsafe { core::mem::transmute(&buffer) };
+            let mut string: String<U26> = String::new();
+            // Little-endian UCS-2 to Big-endian UTF-8
+            for i in 0..data.len() {
+                let ch = unsafe { core::char::from_u32_unchecked(u16::from_le(data[i]) as u32) };
```

Comment:
Could you specify in a comment why do you think this is sound ? Same thing for all the others `unsafe` uses.

Reply:
Modified a bit and explained by variable name. The original code unwrapped the result from `core::char::from_u32`, which leads to panic if it's not a valid rust char, Silently ignore invalid char would be a better choice, otherwise returns Result type for caller to resolve should be another choice.

Reply:
Creating an invalid `char` is UB, which IMO is way worse than a panic.

Reply:
A panic generated by third-party library is also unacceptable, as a third-party library you should not make choice for the user, but give the choise to user, or aware user filename is possibly invalid.
Even if a filename contains invalid char, it's still part of the filename, and we just don't know it's encoding.

Reply:
Then we should return an error, instead of causing undefined behavior.

Reply:
Changed to if not valid char, treat as no LFN.